README.md
# shorts_video_record_button

Our package provides a stylish and modern video recording button that closely resembles the familiar design seen in popular social media platforms like Instagram.


> Instagram-style video recording button

![Example](https://github.com/sooryaraj/shorts_video_record_button/blob/master/assets/screenshots/shorts_video_button_sample_1.gif?raw=true)

![Example 2](https://github.com/sooryaraj/shorts_video_record_button/blob/master/assets/screenshots/shorts_video_button_sample_2.gif?raw=true)

    RecordButton(  
      showLabel: true,  
      labelColor: Colors.black12,  
      trackColor: Colors.grey.shade300,  
      fillColor: Colors.deepOrange,  
      buttonColor: Colors.transparent,  
      gradients: const [  
        Color(0xff405de6),  
        Color(0xff5851db),  
        Color(0xff833ab4),  
        Color(0xffc13584),  
        Color(0xffe1306c),  
        Color(0xfffd1d1d),  
      ],  
      onPlay: () {  
        // Do whatever you want after play
      },  
      onStop: (int value) {  
        // Do whatever you want after stop
      },  
      onComplete: (int value) {  
        // Do whatever you want after complete
      },  
      seconds: 30,  
    );


CHANGELOG.md
## [0.0.1] - 14/08/2023

- Initial release.

## [0.0.2] - 16/08/2023

- Screenshots added


lib/shorts_video_record_button.dart
import 'dart:async';
import 'dart:math';

import 'package:flutter/material.dart';

/// Stylish and modern short video recording button that closely resembles the familiar design seen in popular social media platforms like Instagram.
class RecordButton extends StatefulWidget {
  const RecordButton({
    Key? key,
    this.seconds = 20,
    this.width = 200,
    this.height = 200,
    this.showLabel = false,
    this.labelColor = Colors.blue,
    required this.onPlay,
    required this.onStop,
    required this.onComplete,
    this.fillColor = Colors.green,
    this.trackColor = Colors.black12,
    this.gradients = const [],
    this.strokeCap = StrokeCap.butt,
    this.enable = true,
    this.buttonColor = Colors.red,
  }) : super(key: key);

  final int seconds;
  final double width;
  final double height;
  final bool showLabel;
  final bool enable;
  final Color labelColor;
  final Color buttonColor;
  final Color fillColor;
  final Color trackColor;
  final StrokeCap strokeCap;
  final List<Color> gradients;
  final VoidCallback onPlay;
  final Function(int value) onStop;
  final Function(int value) onComplete;

  @override
  State<RecordButton> createState() => _RecordButtonState();
}

class _RecordButtonState extends State<RecordButton> {
  late _RecordController recordController;

  @override
  void initState() {
    super.initState();
    recordController = _RecordController(seconds: widget.seconds, onComplete: widget.onComplete, onPlay: widget.onPlay, onStop: widget.onStop);
  }

  @override
  Widget build(BuildContext context) {
    return IgnorePointer(
      ignoring: !widget.enable,
      child: Center(
        child: GestureDetector(
          onTap: () {
            recordController.isActive ? recordController.stop(manualStop: true) : recordController.play();
          },
          child: SizedBox(
            width: widget.width,
            height: widget.height,
            child: ListenableBuilder(
                listenable: recordController,
                builder: (context, child) {
                  return CustomPaint(
                      painter: _MyPainter(
                    value: recordController.value.toInt(),
                    seconds: widget.seconds,
                    fillColor: widget.fillColor,
                    trackColor: widget.trackColor,
                    showLabel: widget.showLabel,
                    labelColor: widget.labelColor,
                    gradients: widget.gradients,
                    strokeCap: widget.strokeCap,
                    enable: widget.enable,
                    buttonColor: widget.buttonColor,
                  ));
                }),
          ),
        ),
      ),
    );
  }
}

class _RecordController with ChangeNotifier {
  final int seconds;
  VoidCallback onPlay;
  Function(int value) onStop;
  Function(int value) onComplete;

  _RecordController({required this.seconds, required this.onComplete, required this.onStop, required this.onPlay});

  Timer? _timer;
  int _value = 0;
  bool _isActive = false;

  bool get isActive => _isActive;

  int get value => _value;

  void play() {
    if (_timer != null) return;
    _value = 0;
    onPlay();
    _play();
  }

  void _play() {
    _timer = Timer.periodic(const Duration(milliseconds: 1), (timer) {
      int milliseconds = seconds * 1000;
      if (milliseconds < _value) {
        stop();
      }
      _value = timer.tick;
      notifyListeners();
    });
    _isActive = _timer!.isActive;
    notifyListeners();
  }

  void stop({bool manualStop = false}) {
    _timer?.cancel();
    _timer = null;
    manualStop ? onStop(_value ~/ 1000) : onComplete(_value ~/ 1000);
    _value = 0;
    _isActive = false;
    notifyListeners();
  }
}

class _MyPainter extends CustomPainter {
  _MyPainter({
    required this.value,
    this.trackColor,
    this.strokeCap = StrokeCap.butt,
    this.fillColor,
    this.seconds = 30,
    this.labelColor = Colors.blue,
    this.buttonColor = Colors.red,
    this.showLabel = false,
    this.enable = true,
    this.gradients = const [],
  });

  final int value; // 0 to 10
  final int seconds;
  final Color? trackColor;
  final Color? fillColor;
  final List<Color> gradients;
  final bool showLabel;
  final bool enable;
  final StrokeCap strokeCap;
  final Color labelColor;
  final Color buttonColor;

  @override
  void paint(Canvas canvas, Size size) {
    // Get the center of the canvas
    var bgColor = trackColor ?? Colors.black12;
    var fgColor = fillColor ?? Colors.green;
    var width = size.width;
    var height = size.height;
    final center = Offset(width / 2, height / 2);
    double strokeWidth = width / 10;

    // Draw the gray background seen on the progress indicator
    // This will act as the background layer.

    var innerCirclePaint = Paint();
    if (enable) {
      innerCirclePaint.color = buttonColor;
    } else {
      innerCirclePaint.color = bgColor;
    }
    canvas.drawCircle(
      center,
      (width / 2) - (strokeWidth),
      innerCirclePaint
        ..style = PaintingStyle.fill
        ..strokeWidth = strokeWidth,
    );

    var outerCirclePaint = Paint();
    if (enable) {
      outerCirclePaint.color = bgColor;
    } else {
      outerCirclePaint.color = buttonColor.withOpacity(0.6);
    }
    canvas.drawCircle(
      center,
      (width / 2) - (strokeWidth / 2),
      outerCirclePaint
        ..style = PaintingStyle.stroke
        ..strokeWidth = strokeWidth,
    );

    // Create a new layer where we will be painting the
    // actual progress indicator
    canvas.saveLayer(
      Rect.fromCenter(center: center, width: width, height: height),
      Paint(),
    );

    var rect = Rect.fromCircle(
      center: center,
      radius: 360,
    );

    var fillPaint = Paint();

    if (gradients.isNotEmpty) {
      fillPaint.shader = LinearGradient(
        colors: gradients.reversed.toList(),
      ).createShader(rect);
    } else {
      fillPaint.color = fgColor;
    }
    canvas.drawArc(
      Rect.fromCenter(center: center, width: width - strokeWidth, height: height - strokeWidth),
      radians(0),
      radians(value / (seconds * 1000) * 360),
      false,
      fillPaint
        ..style = PaintingStyle.stroke
        ..strokeCap = strokeCap
        ..strokeWidth = strokeWidth,
    );

    TextSpan span = TextSpan(style: TextStyle(color: labelColor, fontSize: width / 2), text: (value ~/ 1000).toString());
    TextPainter tp = TextPainter(textAlign: TextAlign.center, text: span, textDirection: TextDirection.ltr);
    tp.layout();
    if (showLabel) {
      tp.paint(
        canvas,
        Offset(
          (width - tp.width) * 0.5,
          (height - tp.height) * 0.5,
        ),
      );
    }
    canvas.restore();
  }

  double radians(double degree) {
    return degree * (pi / 180);
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) => true;
}


example/lib/main.dart
import 'dart:developer';

import 'package:flutter/material.dart';
import 'package:shorts_video_record_button/shorts_video_record_button.dart';

void main() {
  runApp(const ShortsRecordButton());
}

class ShortsRecordButton extends StatelessWidget {
  const ShortsRecordButton({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      home: Scaffold(
        body: SizedBox(
          width: MediaQuery.of(context).size.width,
          height: MediaQuery.of(context).size.height,
          child: RecordButton(
            showLabel: true,
            labelColor: Colors.black12,
            trackColor: Colors.grey.shade300,
            fillColor: Colors.deepOrange,
            buttonColor: Colors.transparent,
            gradients: const [
              Color(0xff405de6),
              Color(0xff5851db),
              Color(0xff833ab4),
              Color(0xffc13584),
              Color(0xffe1306c),
              Color(0xfffd1d1d),
            ],
            onPlay: () {
              // Do whatever you want after play
              log("ON PLAY PRESSED=====");
            },
            onStop: (int value) {
              // Do whatever you want after stop
              log("ON STOP PRESSED=====");
            },
            onComplete: (int value) {
              // Do whatever you want after complete
              log("ON COMPLETED===== $value");
            },
            seconds: 3,
          ),
        ),
      ),
    );
  }
}

const backgroundGradient = LinearGradient(
  begin: Alignment.topCenter,
  end: Alignment.bottomCenter,
  colors: [
    Color(0xFFF1D4F5),
    Color(0xFFF9F4FA),
  ],
);

class ShortsRecordSample extends StatelessWidget {
  const ShortsRecordSample({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: SafeArea(
        child: Scaffold(
          body: Container(
            height: MediaQuery.of(context).size.height,
            width: MediaQuery.of(context).size.width,
            decoration: const BoxDecoration(gradient: backgroundGradient),
            child: Stack(
              children: [
                Positioned(
                  bottom: MediaQuery.of(context).size.height * 0.05,
                  child: SizedBox(
                    // height: MediaQuery.of(context).size.height * 0.2,
                    width: MediaQuery.of(context).size.width,
                    child: Column(
                      children: [
                        Row(
                          mainAxisAlignment: MainAxisAlignment.spaceAround,
                          children: [
                            IconButton(
                              icon: const Icon(
                                Icons.camera,
                                color: Colors.black,
                                weight: 30,
                                size: 30,
                              ),
                              onPressed: () {
                                // Handle the button press
                                // toggleCameraDirection();
                              },
                            ),
                            RecordButton(
                                height: 70,
                                width: 70,
                                seconds: 10,
                                showLabel: true,
                                labelColor: Colors.yellow,
                                trackColor: Colors.grey.shade300,
                                // fillColor: Colors.deepOrange,
                                gradients: const [Colors.deepPurple, Colors.pink, Colors.red],
                                onPlay: () {
                                },
                                onStop: (value) {
                                },
                                onComplete: (value) {
                                }),
                            IconButton(
                              icon: const Icon(
                                Icons.cameraswitch,
                                color: Colors.black,
                                weight: 30,
                                size: 30,
                              ),
                              onPressed: () {
                                // toggleCameraDirection();
                              },
                            ),
                          ],
                        ),
                      ],
                    ),
                  ),
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}
